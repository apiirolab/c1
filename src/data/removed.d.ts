type FormControl =
  | 'Input'
  | 'Select'
  | 'Formula'
  | 'Checkbox'
  | 'Radio'
  | 'Textarea'
  | 'Repeater'
  | 'Table'
  | 'Text'
  | 'Form'
  | 'DeleteButton'
  | 'ApproveButton'
  | 'RejectButton';

interface Form {
  name: string;
  description?: string;
  elements?: FormElement[];
}

interface FormElement {
  row?: number;
  column?: number;
  width?: number;
  source?: string;
  label?: string;
  inline?: boolean;
  list?: string;
  filterSource?: string;
  filterColumn?: string;
  control?: FormControl;
  controlProps?: { [index: string]: string };
  vertical?: boolean;
  elements: FormElement[];
}
