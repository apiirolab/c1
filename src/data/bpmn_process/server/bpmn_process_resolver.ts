import { Mutation, purge, Query, Resolver, Yoga } from '../../utils';

export const query: Query = {
  async process(_parent, { id }, ctx, info) {
    return ctx.db.query.bpmnProcess({ where: { id } }, info);
  },
  async processes(
    _parent,
    { input: { status = 'Published', name, skip = 0, first = 20 } },
    ctx,
    info
  ) {
    const user = await ctx.getUser();
    const where = purge<Yoga.BpmnProcessWhereInput>({
      name_contains: name,
      publicationStatus: status,
      OR: [{ readRole: null }, ...user.roles.map(r => ({ readRole_contains: r }))]
    });

    return ctx.db.query.bpmnProcesses(
      {
        orderBy: 'name_ASC',
        where,
        skip,
        first
      },
      info
    );
  }
};

export const mutation: Mutation = {
  async createProcess(_parent, { input: { description, status, name, model } }, ctx, info) {
    const userId = ctx.userId;
    const process = await ctx.db.mutation.createBpmnProcess(
      {
        data: {
          actionCount: 0,
          schema: null,
          resources: null,
          description,
          model,
          name,
          publicationStatus: status,
          type: 'Sales',
          version: 0,
          createdById: userId
        }
      },
      info
    );
    return process;
  }
};
