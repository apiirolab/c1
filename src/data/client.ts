import * as QueryTypes from './generated/types';

export { default as gql } from 'graphql-tag';

export { QueryTypes };

export function purge<T>(params: T): T {
  for (let key of Object.getOwnPropertyNames(params)) {
    if (!(params as any)[key]) {
      delete (params as any)[key];
    }
    if ((params as any)[key] instanceof Object) {
      purge((params as any)[key]);
    }
  }
  return params;
}
