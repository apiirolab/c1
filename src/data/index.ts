export const info = 'Data Info';
