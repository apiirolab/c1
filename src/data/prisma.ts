import * as Prisma from './generated/prisma';

// this is just a helper file that keeps all the types
// it is used for type safety in resolvers
// please add all new business objects here if you want to use them in resolvers

export { Prisma };
