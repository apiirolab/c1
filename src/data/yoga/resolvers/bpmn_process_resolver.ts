import { Query } from '../utils';

export const query: Query = {
  processes(_parent, _args, ctx, info) {
    return ctx.db.query.bpmnProcesses({ where: { status: 'Proposal' } }, info);
  }
};


