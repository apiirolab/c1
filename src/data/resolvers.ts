import * as bpmnProcesses from './bpmn_process/server/bpmn_process_resolver';
import * as bpmnProcessInstances from './bpmn_process_instance/server/process_instance_resolver';
import * as bpmnTaskInstances from './bpmn_task_instance/server/task_instance_resolver';
import * as form from './form/server/form_resolver';
import * as notifications from './notifications/server/notifications_resolver';
import * as resources from './resources/server/resources_resolver';
import * as users from './users/server/user_resolver';

import { extractFragmentReplacements } from 'prisma-binding';

import { addResolvers } from 'data/utils';

/* =========================================================
    Resolvers
   ======================================================== */

export const resolvers = addResolvers([
  notifications,
  bpmnProcesses,
  bpmnProcessInstances,
  bpmnTaskInstances,
  users,
  form,
  resources
]);

export const fragmentReplacements = extractFragmentReplacements(resolvers);
