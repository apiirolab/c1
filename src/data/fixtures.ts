import * as bcrypt from 'bcryptjs';
import { spfBpmn } from './fixtures_data';
import * as Prisma from './generated/prisma';

require('dotenv').config({ path: process.argv[5] });

// tslint:disable-next-line:no-console
console.log('Loading data to: ' + process.env.ENDPOINT);

let db: Prisma.Prisma = new Prisma.Prisma({
  endpoint: process.env.ENDPOINT
  // debug: true
  // secret: 'my_secret123', // only needed if specified in `database/prisma.yml`
});

let loadFixtures = !!process.argv[3];
let createUser = !!process.argv[7];

/* =========================================================
    Constants
   ======================================================== */

const dateStarted = new Date(2018, 1, 2);
const dateFinished = new Date(2018, 2, 10);

const access = (owner: string, role: string, permission: string): Prisma.AccessCreateOneInput => ({
  create: {
    createdById: owner,
    createdOn: dateStarted,
    execute: {
      create: {
        organisationId: null,
        roleId: permission[2] === 'x' && role,
        userId: null
      }
      //#endregion
    },
    read: {
      create: {
        organisationId: null,
        roleId: permission[1] === 'w' && role,
        userId: null
      }
    },
    write: {
      create: {
        organisationId: null,
        roleId: permission[0] === 'r' && role,
        userId: null
      }
    },
    modifiedById: owner,
    modifiedOn: dateFinished
  }
});

/* =========================================================
    DEFAULT DATA
   ======================================================== */

async function loadDefaultLocalisations() {
  // tslint:disable-next-line:no-console
  console.log('Loading default localisations ...');

  await db.mutation.deleteManyLocalisations({});

  await db.mutation.createLocalisation({
    data: { code: 'ProcessStarted', language: 'EN', text: 'Process "{0}" started' }
  });
  await db.mutation.createLocalisation({
    data: { code: 'ProcessFinished', language: 'EN', text: 'Process "{0}" finished' }
  });
  await db.mutation.createLocalisation({
    data: { code: 'ProcessAborted', language: 'EN', text: 'Process "{0}" aborted' }
  });
  await db.mutation.createLocalisation({
    data: { code: 'ActionStarted', language: 'EN', text: 'Action "{0}" started in "{2}"' }
  });
  await db.mutation.createLocalisation({
    data: { code: 'ActionFinished', language: 'EN', text: '{0} completed action "{1}" in "{2}"' }
  });
  await db.mutation.createLocalisation({
    data: {
      code: 'ActionAborted',
      language: 'EN',
      text: '{0} aborted action "{1}" in "{2}" because "{3}"'
    }
  });
  await db.mutation.createLocalisation({
    data: { code: 'ActionRequired', language: 'EN', text: 'Please "{0}" in "{1}" because "{2}"' }
  });
}

// tslint:disable-next-line:no-console
loadDefaultLocalisations().then(() => console.log('Localisations loaded ..'));

export const create = {
  date(addDays: number = 0) {
    return new Date(2018, 2, 10 + addDays);
  },
  userMutation(user: Partial<Prisma.UserCreateInput> = {}) {
    return db.mutation.createUser({ data: create.user(user) });
  },
  user: (user: Partial<Prisma.UserCreateInput> = {}): Prisma.UserCreateInput => ({
    name: 'Tomas Trescak',
    uid: '30031005',
    roles: {
      set: ['admin']
    },
    password: bcrypt.hashSync('123', 10),
    ...user
  }),
  element: (element: Partial<Prisma.FormElementCreateInput>): Prisma.FormElementCreateInput => ({
    column: 0,
    control: 'Input',
    controlProps: null,
    defaultValue: null,
    filterColumn: null,
    filterSource: null,
    inline: false,
    label: null,
    list: null,
    row: 0,
    source: null,
    vertical: null,
    width: null,
    parentElement: null,
    ...element
  }),
  elementMutation(element: Partial<Prisma.FormElementCreateInput> = {}) {
    return db.mutation.createFormElement({ data: create.element(element) });
  },
  validation: (element: Partial<Prisma.ValidatorCreateInput>): Prisma.ValidatorCreateInput => ({
    name: 'RequiredValidator',
    params: null,
    ...element
  }),
  formMutation(form: Partial<Prisma.FormCreateInput> = {}) {
    return db.mutation.createForm({ data: create.form(form) });
  },
  form: (form: Partial<Prisma.FormCreateInput> = {}): Prisma.FormCreateInput => ({
    name: 'Form',
    description: 'Test Form',
    elements: form.elements,
    validations: form.validations,
    ...form
  }),
  descriptor: (
    partial: Partial<Prisma.DataDescriptorCreateInput>
  ): Prisma.DataDescriptorCreateInput => ({
    name: null,
    description: 'Description',
    expression: null,
    type: 'String',
    isArray: false,
    defaultValue: null,
    validators: null,
    parentDescriptor: null,
    ...partial
  }),
  descriptorMutation(partial: Partial<Prisma.DataDescriptorCreateInput>) {
    return db.mutation.createDataDescriptor({ data: create.descriptor(partial) });
  },
  bpmnTask: (partial: Partial<Prisma.BpmnTaskCreateInput>): Prisma.BpmnTaskCreateInput => ({
    taskId: null,
    resources: null,
    name: 'Task Name',
    type: 'Form',
    ...partial
  }),
  bpmnTaskMutation(partial: Partial<Prisma.BpmnTaskCreateInput>) {
    return db.mutation.createBpmnTask({ data: create.bpmnTask(partial) });
  },
  bpmnProcess: (
    partial: Partial<Prisma.BpmnProcessCreateInput>
  ): Prisma.BpmnProcessCreateInput => ({
    name: 'Process',
    description: 'Process 4 description',
    model: '',
    status: 'Published',
    actionCount: 0,
    type: 'HumanResources',
    version: 0,
    access: null,
    resources: null,
    ...partial
  }),
  bpmnProcessMutation(partial: Partial<Prisma.BpmnProcessCreateInput>) {
    return db.mutation.createBpmnProcess({ data: create.bpmnProcess(partial) });
  },
  bpmnProcessInstance: (
    partial: Partial<Prisma.BpmnProcessInstanceCreateInput>
  ): Prisma.BpmnProcessInstanceCreateInput => ({
    process: null,
    data: {},
    owner: null,
    status: 'Running',
    dateStarted,
    ...partial
  }),
  bpmnProcessInstanceMutation(partial: Partial<Prisma.BpmnProcessInstanceCreateInput>) {
    return db.mutation.createBpmnProcessInstance({ data: create.bpmnProcessInstance(partial) });
  },
  bpmnTaskInstance: (
    partial: Partial<Prisma.BpmnTaskInstanceCreateInput>,
    processInstanceId: string,
    taskId: string,
    performerId: string = null
  ): Prisma.BpmnTaskInstanceCreateInput => ({
    dateFinished: null,
    dateStarted,
    duration: 0,
    data: {},
    status: 'Finished',
    performerRoles: { set: ['User'] },
    performer: performerId ? { connect: { id: performerId } } : null,
    processInstance: { connect: { id: processInstanceId } },
    task: { connect: { id: taskId } },
    ...partial
  }),
  bpmnTaskInstanceMutation(
    partial: Partial<Prisma.BpmnTaskInstanceCreateInput>,
    processInstanceId: string,
    taskId: string
  ) {
    return db.mutation.createBpmnTaskInstance({
      data: create.bpmnTaskInstance(partial, processInstanceId, taskId)
    });
  },
  notification: (
    partial: Partial<Prisma.NotificationCreateInput>
  ): Prisma.NotificationCreateInput => ({
    userId: null,
    processInstance: null,
    visible: true,
    code: 'ProcessStarted',
    params: { set: ['Process Name'] },
    type: 'Info',
    ...partial
  }),
  notificationMutation(partial: Partial<Prisma.NotificationCreateInput>) {
    return db.mutation.createNotification({ data: create.notification(partial) });
  },
  resource: (partial: Partial<Prisma.ResourceCreateInput>): Prisma.ResourceCreateInput => ({
    document: null,
    form: null,
    link: null,
    name: 'Resource',
    type: 'Url',
    ...partial
  }),
  resourceMutation(partial: Partial<Prisma.ResourceCreateInput>) {
    return db.mutation.createResource({ data: create.resource(partial) });
  },
  log: (partial: Partial<Prisma.LogCreateInput> = {}): Prisma.LogCreateInput => ({
    date: dateStarted,
    elementId: null,
    elementName: '',
    ...partial
  })
};

/* =========================================================
    FIXTURES
   ======================================================== */

async function insertFixtures() {
  // tslint:disable-next-line:no-console
  console.log('Loading fixtures ...');

  // tslint:disable-next-line:no-console
  console.log('Creating user ...');
  const user = createUser
    ? await create.userMutation({ roles: { set: ['User'] } })
    : (await db.query.users({}))[0];

  // create other user
  const otherUser = createUser
    ? await create.userMutation({ name: 'Other User', uid: '30001234', roles: { set: ['Buyer'] } })
    : (await db.query.users({}))[0];

  /* =========================================================
      Cleanup
     ======================================================== */

  // tslint:disable-next-line:no-console
  console.log('Cleanup ...');

  await db.mutation.deleteManyLogs({});
  await db.mutation.deleteManyDataDescriptors({});
  await db.mutation.deleteManyFormElements({});
  await db.mutation.deleteManyDocuments({});
  await db.mutation.deleteManyForms({});
  await db.mutation.deleteManyResources({});
  await db.mutation.deleteManyNotifications({});
  await db.mutation.deleteManyBpmnTaskInstances({});
  await db.mutation.deleteManyBpmnProcessInstances({});
  await db.mutation.deleteManyBpmnProcesses({});
  await db.mutation.deleteManyBpmnTasks({});

  /* =========================================================
      Resources
     ======================================================== */

  // tslint:disable-next-line:no-console
  console.log('Creating resources ...');

  const nameDescriptor = await create.descriptorMutation({ name: 'owner.personal.name' });
  const ageDescriptor = await create.descriptorMutation({
    name: 'owner.personal.age',
    type: 'Float',
    validators: {
      create: [{ name: 'RequiredValidator' }]
    }
  });
  const childrenDescriptor = await create.descriptorMutation({
    name: 'children',
    type: 'Object',
    isArray: true,
    validators: {
      create: [{ name: 'ArrayLengthValidator', params: { set: ['1'] } }]
    }
  });
  const childDescriptor = await create.descriptorMutation({
    name: 'childName',
    type: 'String',
    validators: {
      create: [{ name: 'RequiredValidator' }]
    },
    parentDescriptor: childrenDescriptor.id
  });
  const addressDescriptor = await create.descriptorMutation({
    name: 'address',
    type: 'Object'
  });
  const cityDescriptor = await create.descriptorMutation({
    name: 'city',
    type: 'String',
    validators: {
      create: [{ name: 'RequiredValidator' }]
    },
    parentDescriptor: addressDescriptor.id
  });
  const zipDescriptor = await create.descriptorMutation({
    name: 'zip',
    type: 'String',
    validators: {
      create: [{ name: 'RegExValidator', params: { set: ['\\d\\d\\d\\d'] } }]
    },
    parentDescriptor: addressDescriptor.id
  });

  const addressElement = await create.elementMutation({
    label: 'Address',
    control: 'Text',
    source: {
      connect: { id: addressDescriptor.id }
    },
    column: 0,
    width: 16,
    row: 1
  });

  const childrenElement = await create.elementMutation({
    label: 'Children',
    control: 'Repeater',
    source: {
      connect: { id: childrenDescriptor.id }
    },
    column: 0,
    width: 16,
    row: 2
  });

  const report = await create.formMutation({
    name: 'Registration Form',
    description: 'This is a registration form',
    elements: {
      create: [
        create.element({
          label: 'Name',
          control: 'Text',
          source: {
            connect: { id: nameDescriptor.id }
          },
          column: 0,
          width: 8,
          row: 0
        }),
        create.element({
          label: 'Age',
          control: 'Text',
          source: {
            connect: { id: ageDescriptor.id }
          },
          column: 8,
          width: 8,
          row: 0
        })
      ]
    }
  });

  const form = await create.formMutation({
    name: 'Registration Form',
    description: 'This is a registration form',
    elements: {
      connect: [{ id: addressElement.id }, { id: childrenElement.id }],
      create: [
        create.element({
          label: 'Name',
          source: {
            connect: { id: nameDescriptor.id }
          },
          column: 0,
          width: 8,
          row: 0
        }),
        create.element({
          label: 'Age',
          source: {
            connect: { id: ageDescriptor.id }
          },
          column: 9,
          width: 8,
          row: 0
        }),
        create.element({
          label: 'City',
          control: 'Text',
          source: {
            connect: { id: cityDescriptor.id }
          },
          parentElement: addressElement.id,
          column: 0,
          width: 8,
          row: 0
        }),
        create.element({
          label: 'Zip',
          control: 'Text',
          source: {
            connect: { id: zipDescriptor.id }
          },
          parentElement: addressElement.id,
          column: 8,
          width: 8,
          row: 0
        }),
        create.element({
          label: 'Name',
          control: 'Text',
          source: {
            connect: { id: childDescriptor.id }
          },
          parentElement: childrenElement.id,
          column: 0,
          width: 8,
          row: 0
        })
      ]
    }
  });

  const formResource = await create.resourceMutation({
    type: 'Form',
    name: 'Approval Form',
    form: {
      connect: { id: form.id }
    }
  });

  const reportResource = await create.resourceMutation({
    type: 'Form',
    name: 'Deans Report',
    form: {
      connect: { id: report.id }
    }
  });

  /* =========================================================
      Tasks
     ======================================================== */

  // tslint:disable-next-line:no-console
  console.log('Creating tasks ...');

  const genericTask = await create.bpmnTaskMutation({ taskId: 'Task_17t05yl' });

  const formTask = await create.bpmnTaskMutation({
    name: 'Form Task',
    taskId: 'Task_14qe8bh',
    resources: {
      connect: {
        id: formResource.id
      }
    },
    type: 'Form'
  });

  const reportTask = await create.bpmnTaskMutation({
    name: 'Report Task',
    taskId: 'Task_0f1st02',
    resources: {
      connect: {
        id: reportResource.id
      }
    },
    type: 'Form'
  });

  /* =========================================================
      BpmnProcesses
     ======================================================== */

  // tslint:disable-next-line:no-console
  console.log('Creating BPMN processes ...');

  const processDefinitions: Prisma.BpmnProcessCreateInput[] = [
    create.bpmnProcess({
      name: 'Process 1',
      description: 'Process 1 description',
      type: 'Purchases',
      dataDescriptors: {
        connect: [
          { id: nameDescriptor.id },
          { id: ageDescriptor.id },
          { id: childrenDescriptor.id },
          { id: childDescriptor.id },
          { id: addressDescriptor.id },
          { id: cityDescriptor.id },
          { id: zipDescriptor.id }
        ]
      },
      tasks: {
        connect: [{ id: formTask.id }, { id: reportTask.id }, { id: genericTask.id }]
      },
      resources: {
        create: [
          {
            type: 'Form',
            name: 'Advanced Form'
          },
          {
            type: 'File',
            name: 'Excel File',
            link: '/uploads/file.doc'
          },
          {
            type: 'Url',
            name: 'External Resource',
            link: 'http://google.com'
          },
          {
            type: 'Document',
            name: 'Guidelines',
            document: {
              create: {
                content: `<h1>Guidelines</h1><p>This is guideline document</p>`,
                title: 'Guidelines'
              }
            }
          }
        ],
        connect: [{ id: formResource.id }, { id: reportResource.id }]
      },
      version: 0,
      model: spfBpmn,
      access: access(user.id, 'User', 'rwx')
    }),
    create.bpmnProcess({
      name: 'Process 4',
      description: 'Process 4 description',
      type: 'HumanResources',
      access: access(user.id, 'User', 'rw-')
    }),
    create.bpmnProcess({
      name: 'Process 2',
      description: 'Process 2 description',
      status: 'Draft',
      type: 'Sales',
      access: access(user.id, 'User', 'r-x')
    }),
    create.bpmnProcess({
      name: 'Process 3',
      description: 'Process 3 description',
      model: '',
      status: 'Proposal',
      actionCount: 0,
      type: 'Travel',
      version: 0,
      access: access(user.id, 'User', 'r--')
    })
  ];

  let processes: Prisma.BpmnProcess[] = [];
  for (let input of processDefinitions) {
    const process = await db.mutation.createBpmnProcess({ data: { ...input } });
    processes.push(process);
  }

  /* =========================================================
        BpmnProcess Instances
       ======================================================== */

  // tslint:disable-next-line:no-console
  console.log('Creating BPMN process instances ...');

  const processInstanceDefinitions: Prisma.BpmnProcessInstanceCreateInput[] = [
    create.bpmnProcessInstance({
      process: { connect: { id: processes[0].id } },
      owner: { connect: { id: user.id } },
      log: {
        create: [
          create.log({
            elementId: 'StartEvent_1',
            elementName: 'Start event',
            date: create.date(0)
          }),
          create.log({ elementId: 'SequenceFlow_1v4vufq', date: create.date(2) }),
          create.log({
            elementId: 'Task_17t05yl',
            date: create.date(2),
            status: 'Started',
            elementName: 'Approve Report'
          }),
          create.log({
            date: create.date(3),
            elementId: 'Task_17t05yl',
            performer: {
              connect: { id: otherUser.id }
            },
            status: 'Approved',
            elementName: 'Approve Report',
            message: 'Everything went well. Bravo!'
          }),
          create.log({ elementId: 'SequenceFlow_0fvr9vw', date: create.date(1) }),
          create.log({ elementId: 'SequenceFlow_1g6j69j' }),
          create.log({
            elementId: 'Task_1c0tszi',
            elementName: 'Submit More',
            date: create.date(4)
          }),
          create.log({ elementId: 'ExclusiveGateway_1r8olns' })
        ]
      }
    }),
    create.bpmnProcessInstance({
      process: { connect: { id: processes[1].id } },
      data: {},
      owner: { connect: { id: user.id } },
      status: 'Running',
      dateStarted
    }),
    create.bpmnProcessInstance({
      process: { connect: { id: processes[0].id } },
      owner: { connect: { id: user.id } },
      status: 'Finished',
      dateFinished
    }),
    create.bpmnProcessInstance({
      process: { connect: { id: processes[0].id } },
      owner: { connect: { id: user.id } },
      status: 'Aborted',
      dateFinished
    }),
    create.bpmnProcessInstance({
      process: { connect: { id: processes[1].id } },
      owner: { connect: { id: user.id } },
      status: 'Finished',
      dateFinished
    })

    // { processId: fixtureContext.processes[1].id }
  ];

  let processInstances: Prisma.BpmnProcessInstance[] = [];
  for (let input of processInstanceDefinitions) {
    const process = await db.mutation.createBpmnProcessInstance({ data: input });
    processInstances.push(process);
  }

  /* =========================================================
      Task Instances
     ======================================================== */

  const pid = processInstances[0].id;
  db.mutation.updateBpmnProcessInstance({
    where: { id: pid },
    data: {
      tasks: {
        create: [
          create.bpmnTaskInstance(
            { dateStarted: create.date(4), dateFinished: create.date(5), status: 'Paused' },
            pid,
            genericTask.id,
            otherUser.id
          ),
          create.bpmnTaskInstance(
            { dateStarted: create.date(2), dateFinished: create.date(3), status: 'Aborted' },
            pid,
            genericTask.id,
            user.id
          ),
          create.bpmnTaskInstance(
            { dateStarted: create.date(0), dateFinished: create.date(1), status: 'Finished' },
            pid,
            formTask.id,
            user.id
          ),
          create.bpmnTaskInstance(
            { dateStarted: create.date(5), dateFinished: create.date(6), status: 'Approved' },
            pid,
            genericTask.id,
            otherUser.id
          ),
          create.bpmnTaskInstance(
            { dateStarted: create.date(8), dateFinished: create.date(9), status: 'Rejected' },
            pid,
            genericTask.id,
            otherUser.id
          ),
          create.bpmnTaskInstance({ dateStarted: create.date(12) }, pid, formTask.id)
        ]
      }
    }
  });

  /* =========================================================
      Log
     ======================================================== */

  /* =========================================================
        Notifications
      ======================================================== */

  // tslint:disable-next-line:no-console
  console.log('Creating notifications ...');

  const notifications: Prisma.NotificationCreateInput[] = [
    create.notification({
      userId: user.id,
      processInstance: { connect: { id: processInstances[0].id } },
      visible: true,
      code: 'ProcessStarted',
      params: { set: ['Process Name'] },
      type: 'Info'
    }),
    create.notification({
      userId: user.id,
      processInstance: { connect: { id: processInstances[0].id } },
      visible: true,
      code: 'ProcessFinished',
      params: { set: ['Process Name'] },
      type: 'Info'
    }),
    create.notification({
      userId: user.id,
      processInstance: { connect: { id: processInstances[0].id } },
      visible: true,
      code: 'ProcessAborted',
      params: { set: ['Process Name'] },
      type: 'Error'
    }),
    create.notification({
      userId: user.id,
      processInstance: { connect: { id: processInstances[0].id } },
      visible: true,
      code: 'ActionStarted',
      params: { set: ['Action Name', 'Process Name'] },
      type: 'Info'
    }),
    create.notification({
      userId: user.id,
      processInstance: { connect: { id: processInstances[0].id } },
      visible: true,
      code: 'ActionFinished',
      params: { set: ['Tomas Trescak', 'Action Name', 'Process Name'] },
      type: 'Info'
    }),
    create.notification({
      userId: user.id,
      processInstance: { connect: { id: processInstances[0].id } },
      visible: true,
      code: 'ActionRequired',
      params: { set: ['Action Name', 'Process Name', 'Tomas Trescak'] },
      type: 'Info'
    }),
    create.notification({
      userId: user.id,
      processInstance: { connect: { id: processInstances[0].id } },
      visible: true,
      code: 'ActionAborted',
      params: { set: ['Tomas Trescak', 'Action Name', 'Process Name', 'Reason'] },
      type: 'Warning'
    })
  ];

  for (let input of notifications) {
    await db.mutation.createNotification({ data: input });
  }

  /* =========================================================
      Cleanup
     ======================================================== */

  // await db.mutation.deleteManyUsers({});
}

if (loadFixtures) {
  // tslint:disable-next-line:no-console
  insertFixtures().then(() => console.log('Fixtures inserted'));
}
