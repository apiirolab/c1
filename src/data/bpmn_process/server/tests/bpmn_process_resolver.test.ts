import { create, its } from 'data/tests';

import { query } from '../bpmn_process_resolver'; 

describe('Bpmn Process Resolver', () => {
  its('fails', { clear: ['BpmnProcessInstance', 'BpmnProcess'] }, async context => {
    await create.process(
      context,
      { status: 'Published', name: 'Process 1' },
      { read: { create: { roleId: 'user' } } }
    );
    await create.process(
      context,
      { status: 'Published', name: 'Process 2' },
      { read: { create: { roleId: 'user' } } }
    );
    await create.process(
      context,
      { status: 'Published', name: 'Process 2' },
      { read: { create: { roleId: 'admin' } } }
    );
    await create.process(
      context,
      { status: 'Draft', name: 'Process 2' },
      { read: { create: { roleId: 'user' } } }
    );
    // mock user
    context.getUser = () => ({ roles: ['user'] } as any);
    // default by role
    // we will get only those that can be read by role 'user'
    let processes = await query.processes(null, { input: {} }, context);
    expect(processes.length).toBe(2);
    // by name
    processes = await query.processes(null, { input: { name: '1' } }, context);
    expect(processes.length).toBe(1);
    processes = await query.processes(null, { input: { name: '2' } }, context);
    expect(processes.length).toBe(1);
    processes = await query.processes(null, { input: { name: '2', status: null } }, context);
    expect(processes.length).toBe(2);
  });
});
