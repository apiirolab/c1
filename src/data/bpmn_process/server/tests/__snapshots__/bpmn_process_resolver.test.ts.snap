// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bpmn Process Resolver fails 1`] = `1`;

exports[`Bpmn Process Resolver tests owner resolver 123 1`] = `
Object {
  "data": Object {
    "bpmnProcessInstanceQuery": Object {
      "comments": Array [
        Object {
          "date": "2018-03-09T13:00:00.000Z",
          "id": "cjnblzn280ztz0b65xxpfjlla",
          "replyTo": null,
          "text": "12345",
          "user": null,
        },
      ],
      "data": "{}",
      "dateFinished": null,
      "dateStarted": "2018-02-01T13:00:00.000Z",
      "id": "cjnblzn280zty0b65pv510t7p",
      "log": Array [],
      "owner": Object {
        "id": "cjnblzn120zts0b65o6soe4ip",
        "name": "Tomi",
      },
      "process": null,
      "status": "Running",
      "tasks": Array [],
    },
  },
}
`;

exports[`Notifications Container renders final 1`] = `
<div>
  <h5
    className="ui dividing header"
  >
    <i
      aria-hidden="true"
      className="bell icon"
    />
    <div
      className="content"
    >
      Notifications
    </div>
  </h5>
  <div
    className="ui divided relaxed list"
    role="list"
  >
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      <i
        aria-hidden="true"
        className="icon"
      >
        <i
          aria-hidden="true"
          className="grey eye icon sc-bwzfXH kZUMPF"
          onClick={[Function]}
          title="Mark as viewed"
        />
      </i>
      <div
        className="content"
      >
        <div>
          <span
            className="sc-bdVaJa cWCCxy"
          >
            2 months ago
          </span>
           ·
           
          <a
            href="/process/process-name"
          >
            Notification Text
          </a>
        </div>
      </div>
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      <i
        aria-hidden="true"
        className="icon"
      >
        <i
          aria-hidden="true"
          className="grey eye icon sc-bwzfXH kZUMPF"
          onClick={[Function]}
          title="Mark as viewed"
        />
      </i>
      <div
        className="content"
      >
        <div>
          <span
            className="sc-bdVaJa cWCCxy"
          >
            2 months ago
          </span>
           ·
           
          <a
            href="/process/process-name"
          >
            Notification Text
          </a>
        </div>
      </div>
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      <i
        aria-hidden="true"
        className="icon"
      >
        <i
          aria-hidden="true"
          className="grey eye icon sc-bwzfXH kZUMPF"
          onClick={[Function]}
          title="Mark as viewed"
        />
      </i>
      <div
        className="content"
      >
        <div>
          <span
            className="sc-bdVaJa cWCCxy"
          >
            2 months ago
          </span>
           ·
           
          <a
            href="/process/process-name"
          >
            Notification Text
          </a>
        </div>
      </div>
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      <i
        aria-hidden="true"
        className="icon"
      >
        <i
          aria-hidden="true"
          className="grey eye icon sc-bwzfXH kZUMPF"
          onClick={[Function]}
          title="Mark as viewed"
        />
      </i>
      <div
        className="content"
      >
        <div>
          <span
            className="sc-bdVaJa cWCCxy"
          >
            2 months ago
          </span>
           ·
           
          <a
            href="/process/process-name"
          >
            Notification Text
          </a>
        </div>
      </div>
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      <i
        aria-hidden="true"
        className="icon"
      >
        <i
          aria-hidden="true"
          className="grey eye icon sc-bwzfXH kZUMPF"
          onClick={[Function]}
          title="Mark as viewed"
        />
      </i>
      <div
        className="content"
      >
        <div>
          <span
            className="sc-bdVaJa cWCCxy"
          >
            2 months ago
          </span>
           ·
           
          <a
            href="/process/process-name"
          >
            Notification Text
          </a>
        </div>
      </div>
    </div>
  </div>
  <button
    className="ui red icon left labeled button"
    onClick={[Function]}
    role="button"
  >
    <i
      aria-hidden="true"
      className="trash icon"
    />
    Clear All
  </button>
</div>
`;

exports[`Notifications Container renders loading 1`] = `
<div>
  Loading...
</div>
`;
